### PR TITLE
`macros`: match unused vars if logging is disabled

### LINKED ISSUE #494: semver-breaking: "Unused variable" errors with deny(unused).
The following used to build fine with defmt 0.2.1, and now fails with defmt 0.2.2. This is due to #477 

```rust
#![deny(unused)]
let x = 1; // error: unused variable: `x` if no "defmt-xxx" features are enabled
info!("x={}", x);
```

### PR DESCRIPTION
**Pull request by Urhengulas**
Add `cfg_if 1.0` to `defmt`.

Fixes #494

### CODE REVIEW

File: macros/src/lib.rs [hunk 1]
```diff
@@ -513,16 +513,22 @@ fn log(level: Level, log: FormatArgs) -> TokenStream2 {
...
     quote!({
-        #[#logging_enabled] {
-            match (#(&(#args)),*) {
-                (#(#pats),*) => {
-                    if let Some(mut _fmt_) = defmt::export::acquire() {
-                        _fmt_.header(&defmt::export::istr(#sym));
-                        #(#exprs;)*
-                        _fmt_.finalize();
-                        defmt::export::release(_fmt_)
+        cfg_if::cfg_if! {
```
Comment (Dirbaio) [inline macros/src/lib.rs:516]: This breaks if the end-user's crate doesn't have cfg_if as a dependency

Comment (Urhengulas) [reply macros/src/lib.rs:516]: Should't it be okay, since I've added it to `defmt`s dependencies?

Comment (Urhengulas) [reply macros/src/lib.rs:516]: Or do you mean if someone uses `defmt-macros` directly?

Comment (Urhengulas) [reply macros/src/lib.rs:516]: I see 🤔

**Comment by Urhengulas** [comment]
Re-exporting solves the problem @Dirbaio brought up, but I am very unsure if we really want to do this. One the one hand it works and also `cfg_if` seems quite stable. On the other hand it feels like a weird solution, but honestly all macro-related things feel a bit weird to me.
Opinions?

On another note: Does anybody know what the dependencies `core` and `compiler_builtins` are used for? Can we just disable them?

**Comment by jonas-schievink** [comment]
Wouldn't it be easier to just avoid `cfg_if` and use the negated `#[cfg]` expression?

**Comment by Urhengulas** [comment]
> Wouldn't it be easier to just avoid `cfg_if` and use the negated `#[cfg]` expression?

Yes, but then we need to maintain two config-expressions and change the other one, in case the first one does. `cfg_if` seemed like the cleaner and more readable one to me.

On the other hand we probably won't change it too often, so it is okay if we add a comment warning about this relationship.

In the end I am fine with both. What do you prefer?

**Comment by jonas-schievink** [comment]
The condition is generated by the macro (`#[#logging_enabled]`), inverting that should just be `#[not(#logging_enabled)]`

**Comment by Urhengulas** [comment]
> The condition is generated by the macro (`#[#logging_enabled]`), inverting that should just be `#[not(#logging_enabled)]`

Oh, that's nice. I though that the `not` (and others) only work inside the `cfg(...)`, but good to know it is more flexible.

**Comment by jonas-schievink** [comment]
Err, no, you're right, it has to be `#[cfg(not(...))]`, but a slight refactoring should let you do that

**Comment by Urhengulas** [comment]
> Err, no, you're right, it has to be `#[cfg(not(...))]`, but a slight refactoring should let you do that

Also just noticed xD `error[E0658]: attributes on expressions are experimental`

**Pull request closed without merging**